File: dfs/paths/binary_tree_max_path_sum.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# @param {TreeNode} root
# @return {Integer}
def max_path_sum(root)
    @max = -Float::INFINITY
    dfs(root)

    @max
end

def dfs(root)
    return 0 if root.nil?
    l, r = dfs(root.left), dfs(root.right)
    cur = l + root.val + r # Path sum
    @max = [cur, @max].max

    [l + root.val, r + root.val, 0].max # Node sum
end

# Things to keep track of
# 1. Node Sum: Maximum path sum ending at the current node (Returned from dfs)
# 2. Path Sum: Path sum for path including current node (Update max if this is greater)

# 124. Binary Tree Maximum Path Sum
# https://leetcode.com/problems/binary-tree-maximum-path-sum/description/

require 'test/unit'
extend Test::Unit::Assertions

left = TreeNode.new(2)
right = TreeNode.new(3)
root = TreeNode.new(1)
root.left, root.right = left, right
assert_equal(max_path_sum(root), 6)

l2_left2, l2_right2 = TreeNode.new(15), TreeNode.new(7)
l1_left, l1_right = TreeNode.new(9), TreeNode.new(20)
l1_right.left, l1_right.right = l2_left2, l2_right2
root = TreeNode.new(-10)
root.left, root.right = l1_left, l1_right
assert_equal(max_path_sum(root), 42)

File: dfs/paths/binary_tree_max_path_sum.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        self.max_sum = -float('inf')
        self.dfs(root)
        return self.max_sum

    def dfs(self, node: TreeNode) -> None:
        if not node: return 0

        l, r = self.dfs(node.left), self.dfs(node.right)
        self.max_sum = max(self.max_sum, l + node.val + r, l + node.val, r + node.val, node.val)

        return max(0, max(l, r) + node.val, node.val)


# 124. Binary Tree Maximum Path Sum
# https://leetcode.com/problems/binary-tree-maximum-path-sum/description/


# Exactly 
# Intuition:
# Things to keep track of:
# 1. Node Sum: Maximum path sum ending at the current node (Returned from dfs)
# 2. Path Sum: Path sum for path including current node (Update max if this is greater)


left = TreeNode(2)
right = TreeNode(3)
root = TreeNode(1)
root.left, root.right = left, right
sol = Solution()
assert sol.maxPathSum(root) == 6

l2_left2, l2_right2 = TreeNode(15), TreeNode(7)
l1_left, l1_right = TreeNode(9), TreeNode(20)
l1_right.left, l1_right.right = l2_left2, l2_right2
root = TreeNode(-10)
root.left, root.right = l1_left, l1_right
assert sol.maxPathSum(root) == 42
